xds/internal/balancer/priority/balancer.go

```diff
@@ -0,0 +1,221 @@
+	//
+	// One timer is enough because there can be at most one priority in init
+	// state.
+	priorityInitTimer *time.Timer
+}
+
+func newPriorityBalancer(cc balancer.ClientConn) *priorityBalancer {
+	b := &priorityBalancer{
+		cc:               cc,
+		childToPriority:  make(map[string]int),
+		children:         make(map[string]*childBalancer),
+		childStateUpdate: buffer.NewUnbounded(),
+	}
+	b.ctx, b.cancel = context.WithCancel(context.Background())
+
+	b.logger = prefixLogger(b)
+	b.bg = balancergroup.New(cc, b, nil, b.logger)
+	b.bg.Start()
+	go b.run()
+	b.logger.Infof("Created")
+	return b
+}
+
+func (pb *priorityBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	newConfig, ok := s.BalancerConfig.(*lbConfig)
+	if !ok {
+		return fmt.Errorf("unexpected balancer config with type: %T", s.BalancerConfig)
+	}
+	addressesSplit := hierarchy.Group(s.ResolverState.Addresses)
+
+	pb.mu.Lock()
+	defer pb.mu.Unlock()
+	// Create and remove children, since we know all children from the config
+	// are used by some priority.
+	for name, newSubConfig := range newConfig.Children {
+		bb := balancer.Get(newSubConfig.Config.Name)
+		if bb == nil {
+			pb.logger.Warningf("balancer name %v from config is not registered", newSubConfig.Config.Name)
+			continue
+		}
+
+		currentChild, ok := pb.children[name]
+		if !ok {
+			// This is a new child, add it to the children list. But note that
+			// the balancer isn't built, because this child can be a low
+			// priority. If necessary, it will be built when syncing priorities.
+			cb := newChildBalancer(name, pb, bb)
+			cb.updateConfig(newSubConfig.Config.Config, resolver.State{
+				Addresses:     addressesSplit[name],
+				ServiceConfig: s.ResolverState.ServiceConfig,
+				Attributes:    s.ResolverState.Attributes,
+			})
+			pb.children[name] = cb
+			continue
+		}
+
+		// This is not a new child. But the config/addresses could change.
+
+		// The balancing policy name is changed, close the old child. But don't
+		// rebuild, rebuild will happen when syncing priorities.
+		if currentChild.bb.Name() != bb.Name() {
+			currentChild.stop()
+			currentChild.bb = bb
+		}
+
+		// Update config and address, but note that this doesn't send the
+		// updates to child balancer (the child balancer might not be built, if
+		// it's a low priority).
+		currentChild.updateConfig(newSubConfig.Config.Config, resolver.State{
+			Addresses:     addressesSplit[name],
+			ServiceConfig: s.ResolverState.ServiceConfig,
+			Attributes:    s.ResolverState.Attributes,
+		})
+	}
+	// Remove child from children if it's not in new config.
+	for name, oldChild := range pb.children {
+		if _, ok := newConfig.Children[name]; !ok {
+			oldChild.stop()
+		}
+	}
+
+	// Update priorities and handle priority changes.
+	pb.priorities = make([]string, 0, len(newConfig.Priorities))
+	pb.childToPriority = make(map[string]int)
+	for pi, pName := range newConfig.Priorities {
+		pb.priorities = append(pb.priorities, pName)
```

Comment:
Can't we directly assign `pb.priorities = newConfig.Priorities`? If not, can we say `pb.priorities[pi] = pName` since you have already made the slice as big as required.

Reply:
Hmm, yeah, directly assigning works! Done